File: app/src/test/java/me/riek/QuestionsTest.kt
package me.riek

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.abs
import kotlin.random.Random

class QuestionsTest {

    private fun operand(s: String): Double =
        if ("/" in s) { val (n, d) = s.split("/"); n.toDouble() / d.toDouble() } else s.toDouble()

    private fun evalText(text: String): Double {
        val (a, op, b) = text.split(" ")
        val x = operand(a); val y = operand(b)
        return when (op) {
            "+" -> x + y
            "−" -> x - y
            "×" -> x * y
            "÷" -> x / y
            else -> error("unknown op '$op' in $text")
        }
    }

    @Test fun generatedQuestionsMatchTheSite() {
        repeat(500) { seed ->
            val qs = generateQuestions(Random(seed))
            assertEquals(QUESTION_COUNT, qs.size)
            qs.forEach { q ->
                // the printed equation must actually evaluate to the answer
                assertTrue(
                    "equation ${q.text} != ${q.answer}",
                    abs(evalText(q.text) - operand(q.answer)) < 1e-4,
                )
                // typing the exact answer must score correct (site's checker)
                assertTrue("round-trip failed for ${q.answer}", checkAnswer(q.answer, q.answer))
            }
        }
    }

    @Test fun answerCheckerBehavesLikeTheSite() {
        assertTrue(checkAnswer(" 42 ", "42"))          // trim + exact
        assertTrue(checkAnswer("5/6", "5/6"))          // fraction exact
        assertTrue(checkAnswer("12.50", "12.5"))       // numeric within 1e-4
        assertTrue(!checkAnswer("", "7"))              // blank is never correct
        assertTrue(!checkAnswer("0.83", "5/6"))        // decimal != fraction string
    }
}

File: web/src/lib/questions.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { generateQuestions, checkAnswer, QUESTION_COUNT } from "./questions.ts";

const operand = (s: string): number =>
  s.includes("/") ? Number(s.split("/")[0]) / Number(s.split("/")[1]) : Number(s);

const evalText = (text: string): number => {
  const [a, op, b] = text.split(" ");
  const x = operand(a), y = operand(b);
  switch (op) {
    case "+": return x + y;
    case "−": return x - y;
    case "×": return x * y;
    case "÷": return x / y;
    default: throw new Error(`unknown op '${op}' in ${text}`);
  }
};

test("generated questions match their printed equation and round-trip", () => {
  for (let i = 0; i < 500; i++) {
    const qs = generateQuestions();
    assert.equal(qs.length, QUESTION_COUNT);
    for (const q of qs) {
      assert.ok(Math.abs(evalText(q.text) - operand(q.answer)) < 1e-4, `equation ${q.text} != ${q.answer}`);
      assert.ok(checkAnswer(q.answer, q.answer), `round-trip failed for ${q.answer}`);
    }
  }
});

test("answer checker behaves like the site", () => {
  assert.ok(checkAnswer(" 42 ", "42"));
  assert.ok(checkAnswer("5/6", "5/6"));
  assert.ok(checkAnswer("12.50", "12.5"));
  assert.ok(!checkAnswer("", "7"));
  assert.ok(!checkAnswer("0.83", "5/6"));
});
